### PR TITLE
Support thinking budget for Inkling

### PR DESCRIPTION
**Pull request by draftbk**
## Motivation

Inkling has no thinking-budget processor. Adding one surfaced a base-class bug: a thinking block counts as closed if the end token appears anywhere in `prompt + output`, and Inkling's end token also terminates every message, so `thinking_budget` was silently ignored.

## Modifications

- Add `InklingThinkingBudgetLogitProcessor`.
- Resolve the open thinking block by scanning back to the most recent delimiter. Unchanged for a sequence with one delimiter, the normal shape for GLM-4.5 / Qwen3 / DeepSeek-R1; existing tests pass unmodified.
- Rename and re-budget `test_multiple_thinking_start_counts_from_first`, which passed under either semantics.

## Accuracy Tests

Served `Inkling-Small-NVFP4` on 2×B200:

```bash
python -m sglang.launch_server --model-path <path>/Inkling-Small-NVFP4 \
  --tp 2 --quantization modelopt_fp4 --attention-backend fa4 \
  --fp4-gemm-backend flashinfer_trtllm --moe-runner-backend flashinfer_trtllm_routed \
  --page-size 128 --mem-fraction-static 0.85 --reasoning-parser inkling \
  --enable-custom-logit-processor
```

Same prompt 8 times per arm, at `reasoning_effort: high` so the model wants to think for hundreds of tokens:

```python
import openai
from sglang.srt.sampling.custom_logit_processor import InklingThinkingBudgetLogitProcessor

client = openai.Client(base_url="http://127.0.0.1:30000/v1", api_key="*")
response = client.chat.completions.create(
    model=MODEL,
    messages=[{"role": "user", "content":
               "A bat and a ball cost $1.10 in total. The bat costs $1.00 more "
               "than the ball. How much does the ball cost?"}],
    max_tokens=2048,
    extra_body={
        "chat_template_kwargs": {"reasoning_effort": "high"},
        "custom_logit_processor": InklingThinkingBudgetLogitProcessor().to_str(),
        "custom_params": {"thinking_budget": 60},
    },
)
```

Thinking tokens per response, from tokenizing `reasoning_content`:

| arm | counts | over budget |
|---|---|---|
| no budget set | 74, 111, 119, 122, 141, 179, 227, 237 | 8/8 |
| `main`'s logic | 105, 134, 148, 148, 154, 176, 177, 177 | 8/8 |
| **this PR** | **33, 47, 51, 61, 62, 62, 62, 62** | **0/8** |

On `main` the budget has no effect. With the fix, nothing exceeds it: responses stop at the budget plus the one or two tokens spent forcing a newline and then the end token, or earlier when the model finishes thinking on its own. The answer was correct in all 24 responses. The middle arm ships `main`'s `__call__` through the same mechanism, since the processor class does not exist there.

Unit tests: 35 pass. Against unpatched `main`, `test_end_token_in_prompt_does_not_disable_budget` and `test_multiple_thinking_start_counts_from_most_recent` fail.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #31356491674](https://github.com/sgl-project/sglang/actions/runs/31356491674)
Latest PR Test (Extra): :x: [Run #31356491464](https://github.com/sgl-project/sglang/actions/runs/31356491464)

### CODE REVIEW

**Comment by draftbk** [comment]
Hi @JustinTong0323, tagging you since you reviewed #11416. Would appreciate your review, and happy to change anything.

File: test/registered/unit/sampling/test_custom_logit_processor.py [hunk 1]
```diff
@@ -224,21 +224,34 @@ def test_none_param_dict_in_list_skipped(self):
...
-        params = [{"thinking_budget": 10, "__req__": req}]
+        # Most recent START at index 3, tokens_after_start = 2
+        # Budget=4 → 2 < 4 → no modification (index 0 would have given 5)
+        params = [{"thinking_budget": 4, "__req__": req}]
         logits = self._logits()
         original = logits.clone()
         result = self.processor(logits, params)
         self.assertTrue(torch.equal(result, original))
 
+    def test_end_token_in_prompt_does_not_disable_budget(self):
```
Comment (JustinTong0323) [inline test/registered/unit/sampling/test_custom_logit_processor.py]: Exercise `InklingThinkingBudgetLogitProcessor` directly here, assert its framing IDs against `inkling_tokenizer.py`, and cover the forced newline/end path; the current Qwen3-based test leaves the new Inkling constants untested.

**Review by JustinTong0323 — COMMENTED** [review]
The delimiter fix looks correct; add direct coverage for the new Inkling subclass.

File: test/registered/unit/sampling/test_custom_logit_processor.py (hunk 1 above)
Comment (JustinTong0323) [reply test/registered/unit/sampling/test_custom_logit_processor.py]: Fixed directly in 563c6672bb. The regression now uses `InklingThinkingBudgetLogitProcessor`, checks its framing IDs against `inkling_tokenizer.py`, and covers both the forced newline and end-token steps.

**Review by JustinTong0323 — APPROVED** [review]
LGTM
